session.go

```diff
@@ -128,61 +150,50 @@ func (s *session) RemoteHeader() http.Header {
 }
 
 func (s *session) nextReader() (base.FrameType, base.PacketType, io.ReadCloser, error) {
-	var ft base.FrameType
-	var pt base.PacketType
-	var r io.ReadCloser
-	var err error
 	for {
 		s.upgradeLocker.RLock()
-		ft, pt, r, err = s.conn.NextReader()
+		conn := s.conn
+		s.upgradeLocker.RUnlock()
+		ft, pt, r, err := conn.NextReader()
 		if err != nil {
-			s.upgradeLocker.RUnlock()
-			if op, ok := err.(payload.Error); ok {
-				if op.Temporary() {
-					continue
-				}
+			if op, ok := err.(payload.Error); ok && op.Temporary() {
+				continue
 			}
 			return 0, 0, nil, err
 		}
-		s.upgradeLocker.RUnlock()
-		return ft, pt, newReader(r, &s.upgradeLocker), nil
```

Comment:
Note: Using `upgradeLocker` for the reader's lock was an improper (re)use of the `upgradeLocker` mutex which appears to be intended to only protect the `conn` and `transport` fields of the `session` struct, rather than also protect concurrent writes on the websocket connection.

But this is irrelevant since the `reader`/`newReader` stuff in util.go is unnecessary given `transport/websocket.wrapper`.